File: AtCoder/ABC/abc373/e/main.nim
import std/times
import std/random
# import nimprof
{.checks: off.}
import macros;macro ImportExpand(s:untyped):untyped = parseStmt($s[2])
# https://atcoder.jp/users/kemuniku さんのマクロを勝手に借りてます。
ImportExpand "cplib/tmpl/sheep.nim" <=== "when not declared CPLIB_TMPL_SHEEP:\n    const CPLIB_TMPL_SHEEP* = 1\n    {.warning[UnusedImport]: off.}\n    {.hint[XDeclaredButNotUsed]: off.}\n    import algorithm\n    import sequtils\n    import tables\n    import macros\n    import math\n    import sets\n    import strutils\n    import strformat\n    import sugar\n    import heapqueue\n    import streams\n    import deques\n    import bitops\n    import std/lenientops\n    import options\n    #入力系\n    proc scanf(formatstr: cstring){.header: \"<stdio.h>\", varargs.}\n    proc getchar(): char {.importc: \"getchar_unlocked\", header: \"<stdio.h>\", discardable.}\n    proc ii(): int {.inline.} = scanf(\"%lld\\n\", addr result)\n    proc lii(N: int): seq[int] {.inline.} = newSeqWith(N, ii())\n    proc si(): string {.inline.} =\n        result = \"\"\n        var c: char\n        while true:\n            c = getchar()\n            if c == ' ' or c == '\\n':\n                break\n            result &= c\n    #chmin,chmax\n    template `max=`(x, y) = x = max(x, y)\n    template `min=`(x, y) = x = min(x, y)\n    #bit演算\n    proc `%`*(x: int, y: int): int =\n        result = x mod y\n        if y > 0 and result < 0: result += y\n        if y < 0 and result > 0: result += y\n    proc `//`*(x: int, y: int): int{.inline.} =\n        result = x div y\n        if y > 0 and result * y > x: result -= 1\n        if y < 0 and result * y < x: result -= 1\n    proc `%=`(x: var int, y: int): void = x = x%y\n    proc `//=`(x: var int, y: int): void = x = x//y\n    proc `**`(x: int, y: int): int = x^y\n    proc `**=`(x: var int, y: int): void = x = x^y\n    proc `^`(x: int, y: int): int = x xor y\n    proc `|`(x: int, y: int): int = x or y\n    proc `&`(x: int, y: int): int = x and y\n    proc `>>`(x: int, y: int): int = x shr y\n    proc `<<`(x: int, y: int): int = x shl y\n    proc `~`(x: int): int = not x\n    proc `^=`(x: var int, y: int): void = x = x ^ y\n    proc `&=`(x: var int, y: int): void = x = x & y\n    proc `|=`(x: var int, y: int): void = x = x | y\n    proc `>>=`(x: var int, y: int): void = x = x >> y\n    proc `<<=`(x: var int, y: int): void = x = x << y\n    proc `[]`(x: int, n: int): bool = (x and (1 shl n)) != 0\n    #便利な変換\n    proc `!`(x: char, a = '0'): int = int(x)-int(a)\n    #定数\n    #[ include cplib/utils/constants ]#\n    when not declared CPLIB_UTILS_CONSTANTS:\n        const CPLIB_UTILS_CONSTANTS* = 1\n        const INF32*: int32 = 100100111.int32\n        const INF64*: int = int(3300300300300300491)\n    const INF = INF64\n    #converter\n\n    #range\n    iterator range(start: int, ends: int, step: int): int =\n        var i = start\n        if step < 0:\n            while i > ends:\n                yield i\n                i += step\n        elif step > 0:\n            while i < ends:\n                yield i\n                i += step\n    iterator range(ends: int): int = (for i in 0..<ends: yield i)\n    iterator range(start: int, ends: int): int = (for i in\n            start..<ends: yield i)\n\n    #joinが非stringでめちゃくちゃ遅いやつのパッチ\n    proc join*[T: not string](a: openArray[T], sep: string = \"\"): string = a.mapit($it).join(sep)\n"  

var 
    n,m,k = ii()
    a = lii(n)
    sa = sorted(a)
    pa = newSeq[int](n+1)
for i in 0..<n:
    pa[i+1] = pa[i] + sa[i]
var sum_a = sum(a)
var ans = newSeqWith(n,-2)

for i in 0..<n:
    proc isOK(mid:int):bool=
        var p = n - sa.upperBound(a[i] + mid)
        var le = n - p
        proc isOK2(mid2:int):bool=
            if sa[le - mid2 - 1] < a[i]:
                return mid2 * (a[i] + mid + 1) - (pa[le] - pa[le - mid2 - 1] - a[i]) <= k - sum_a - mid
            else:
                return mid2 * (a[i] + mid + 1) - (pa[le] - pa[le - mid2]) <= k - sum_a - mid
        proc meguru2(ng,ok:var int):int=
            while abs(ok - ng) > 1:
                var mid = (ok + ng) // 2
                if isOK2(mid):
                    ok = mid
                else:
                    ng = mid
            return ok
        var ok = 0
        var ng = le
        p += meguru2(ng, ok)
        return p < m

    proc meguru(ng,ok: var int):int=
        while abs(ok - ng) > 1:
            var mid = (ok + ng) // 2
            if isOK(mid):
                ok = mid
            else:
                ng = mid
        return ok    
    var ng = -1
    var ok = k - sum_a + 1
    var res = meguru(ng,ok)
    if res > k - sum_a:
        res = -1
    ans[i] = res

echo(ans.join(" "))

File: AtCoder/ABC/abc373/e/main.py
from sys import stdin, setrecursionlimit
from itertools import accumulate
from bisect import bisect_left, bisect_right


DEBUG = False
# import pypyjit
# pypyjit.set_param("max_unroll_recursion=-1")
# 外部ライブラリ
# from sortedcontainers import SortedSet, SortedList, SortedDict
setrecursionlimit(10**7)
around4 = ((-1, 0), (1, 0), (0, -1), (0, 1))  # 上下左右
around8 = ((-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1))
inf = float("inf")
mod = 998244353
input = lambda: stdin.readline().rstrip()
pritn = lambda *x: print(*x)
deb = lambda *x: print(*x) if DEBUG else None
PY = lambda: print("Yes")
PN = lambda: print("No")
SI = lambda: input()
IS = lambda: input().split()
II = lambda: int(input())
MII = lambda: map(int, input().split())
LMII = lambda: list(map(int, input().split()))


n, m, k = MII()
a = LMII()
sa = sorted(a)
pa = [0] + list(accumulate(sa))
sum_a = sum(a)
ans = [-2] * n
for i in range(n):

    def isOK(mid):
        # 票数が a[i] + mid を元々超えている人の人数
        p = n - bisect_right(sa, a[i] + mid)
        # 票数が a[i] + mid 以下の人数
        le = n - p

        def isOK2(mid2):
            # 含まれる
            assert le - mid2 - 1 >= 0
            if sa[le - mid2 - 1] < a[i]:
                return (
                    mid2 * (a[i] + mid + 1) - (pa[le] - pa[le - mid2 - 1] - a[i])
                    <= k - sum_a - mid
                )
            # 含まれない
            else:
                return (
                    mid2 * (a[i] + mid + 1) - (pa[le] - pa[le - mid2])
                    <= k - sum_a - mid
                )

        def meguru2(ng, ok):
            while abs(ok - ng) > 1:
                mid = (ok + ng) // 2
                if isOK2(mid):
                    ok = mid
                else:
                    ng = mid
            return ok

        # 自由な票が、k - sum_a - mid のとき、票数が a[i] + mid より大きい人の人数の最大値
        p += meguru2(le, 0)
        return p < m

    def meguru(ng, ok):
        while abs(ok - ng) > 1:
            mid = (ok + ng) // 2
            if isOK(mid):
                ok = mid
            else:
                ng = mid
        return ok

    res = meguru(-1, k - sum_a + 1)
    if res > k - sum_a:
        res = -1
    ans[i] = res
pritn(*ans)
